Fixed mungegithub templatized deployment scheme.

cjwagner opened the pull request:
Fixed problems with template scheme used for deploying the submit-queue. Specifically an empty secret is uploaded for any deployment that doesn't specify a gitthub webhook secret (such as the docs submit-queue) so that it can still match the deployment template.
I also improved the template scheme by adding a template for the token secrets, the volumes and volume claims, and the services. It should be easy to switch to using helm charts now.

I also improved mungegithub's Makefile by adding targets for the webhook secret, volume and volume claim creation, service creation, and secret deployment.

The most useful new target is `first_time_deploy` which lets you deploy  any mungegithub app to a cluster that is empty except for the required persistent disk. The target creates all of the required secrets, volumes, volume claims, services, and config maps, and then creates the actual deployment.

cjwagner commented:
/cc @rmmh @apelisse

rmmh commented:
I don't like having 98% identical 300 line YAML files. It makes it very easy to accidentally break things when you only change one copy. I don't want to be in the situation where a code change and updated kubernetes deployment means the next deploy for contrib will break.

spxtr commented:
> I don't want to be in the situation where a code change and updated kubernetes deployment means the next deploy for contrib will break.

How does what we have now prevent that?
I count ... 43 flags to mungegithub! That's amazing! It's become a huge problem with kubernetes binaries themselves too, that they're configured with a myriad of command-line flags instead of configuration files.

apelisse commented:
First of all, thank you so much @cjwagner for providing a thorough description of what this change does.

> How does what we have now prevent that?

Yeah, at this point most of the pod configuration is argument forwarding anyway. This file is right at the point where no available solution make sense. Either we have to duplicate the mungegithub pod logic, either we have to create unnecessary boilerplate. I won't lose sleep either way. I'll lose sleep over how we should fix kubernetes to handle this type of use-case though.

> I count ... 43 flags to mungegithub! That's amazing! It's become a huge problem with kubernetes binaries themselves too, that they're configured with a myriad of command-line flags instead of configuration files.

I don't really think this is a problem.

apelisse commented:
do you have an opinion @foxish ?

foxish commented:
I am not in favor of this change. 
If we are changing things, we should just use helm. Helm gives us a lot of this templating logic for free, without having to do our own layer on top.

foxish commented:
Created https://github.com/kubernetes/test-infra/issues/3200

cjwagner commented:
I updated the commits. Now instead of removing the templatization scheme, I fixed it and made templates for more resource specs files that were almost identical. It should be pretty easy to switch to using helm charts from this scheme.

apelisse commented:
I think this is ok.

spiffxp commented on mungegithub/README.md:
````diff
@@ -100,15 +103,18 @@ The steps below make use of the utility cluster which runs the existing submit-q
+* A kubernetes secret named `<TARGET>-github-secret` may be required if the submit-queue is configured to accept github webhooks.  If this is the case then create it by puting the webhook secret into the file `mungegithub/hook-secret` and using the command below.
 ```
-make secret APP=submit-queue TARGET=<TARGET>
+APP=submit-queue TARGET=<TARGET> make push_hook_secret
````
push_hook_secret doesn't exist?

cjwagner replied:
I forgot to update that part of the README. Fixed.

spiffxp commented on mungegithub/submit-queue/.gitignore:
```diff
@@ -1,3 +1,5 @@
 local.secret.yaml
 local.service.yaml
+local.pvc.yaml
+local.pv.yaml
```
there could be less copy-pasta amongst the various .gitignores if they're each `local.*.yaml`, unless the intent is to use this specificity as an extra safeguard

cjwagner replied:
I think its actually safer to use the `local.*.yaml` version. If a secret, service, or persistent volume template is added to an app this will prevent a new local.XXX.yaml file from being committed without requiring an update to .gitignore. Any files that are intended to be committed wont match that pattern.

spiffxp commented:
this is awesome work, my only nit is the non-existent make target

rmmh commented on mungegithub/Makefile:
```diff
@@ -64,14 +80,42 @@ ifeq ($(READONLY),false)
-	@echo $(token64)
-	sed -e 's|1234567890123456789012345678901234567890123456789012345=|$(token64)|' $(APP)/deployment/$(TARGET)/secret.yaml > $(APP)/local.secret.yaml
+	# Updating secret.yaml and outputing to $(APP)/local.secret.yaml.
+	sed -e 's|token1234567890123456789012345678901234567890123456789012345=|$(token64)|' $(APP)/secret.yaml > $(APP)/local.secret.yaml
```
why not just "TOKEN" ?

cjwagner replied:
fixed.

rmmh commented on mungegithub/.gitignore:
```diff
@@ -1 +1,3 @@
 mungegithub
+token
+hook-secret
```
if you add `local.*.yaml` here, you can delete the other .gitignores.

cjwagner replied:
fixed.

rmmh commented:
/lgtm
